### DISCUSSION TITLE
oracledb.exceptions.DatabaseError: Error while trying to retrieve text for error ORA-01804

### DISCUSSION THREAD

**Discussion by kcjob** [oracle/python-oracledb#344] [Q&A]
I'm using the following code on a Mac Sanoma, trying to connect to a remote Oracle database version: 12.1.0.2.0
I don't have any environment variables set

if platform.system() == "Darwin" and platform.machine() == "x86_64":   # macOS
    d = r"/Users/xxxx/oracle/instantclient_19_16"
    oracledb.init_oracle_client(lib_dir=d)    
elif platform.system() == "Windows":
    d = r"/Users/xxxx/oracle/instantclient_21_14"
    oracledb.init_oracle_client(lib_dir=d)    
--------------------------------------------------------

I'm getting to following error:

ODPI [5061619] 2024-06-11 18:48:16.843: ODPI-C 4.3.0
ODPI [5061619] 2024-06-11 18:48:16.843: debugging messages initialized at level 64
ODPI [5061619] 2024-06-11 18:48:21.429: ODPI-C 5.2.0
ODPI [5061619] 2024-06-11 18:48:21.429: debugging messages initialized at level 64
ODPI [5061619] 2024-06-11 18:48:21.429: Context Parameters:
ODPI [5061619] 2024-06-11 18:48:21.429:     Oracle Client Lib Dir: /Users/xxxx/oracle/instantclient_19_16
ODPI [5061619] 2024-06-11 18:48:21.429: Environment Variables:
ODPI [5061619] 2024-06-11 18:48:21.429:     LD_LIBRARY_PATH => "/Users/xxxx/oracle/instantclient_19_16"
ODPI [5061619] 2024-06-11 18:48:21.429: load in parameter directory
ODPI [5061619] 2024-06-11 18:48:21.429: load in dir /Users/xxxx/oracle/instantclient_19_16
ODPI [5061619] 2024-06-11 18:48:21.429: load with name /Users/xxxx/oracle/instantclient_19_16/libclntsh.dylib
ODPI [5061619] 2024-06-11 18:48:22.765: load by OS successful
ODPI [5061619] 2024-06-11 18:48:22.765: validating loaded library
Traceback (most recent call last):
  File "import_dsd_data.py", line 22, in <module>
    oracledb.init_oracle_client(lib_dir=d)    
  File "src/oracledb/impl/thick/utils.pyx", line 495, in oracledb.thick_impl.init_oracle_client
  File "src/oracledb/impl/thick/utils.pyx", line 521, in oracledb.thick_impl.init_oracle_client
  File "src/oracledb/impl/thick/utils.pyx", line 446, in oracledb.thick_impl._raise_from_info
oracledb.exceptions.DatabaseError: Error while trying to retrieve text for error ORA-01804

I've googled everything and can't find an answer

**Comment by cjbj** [comment]
What's the exact code you are running? Give a runnable script so we know exactly and don't have to guess. (Please cut out any Windows etc code!)

I wouldn't expect to see LD_LIBRARY_PATH in the trace unless you had it set. Unset it to avoid confusion.  

Check that you really don't have any Oracle environment variables set.

Check file permissions on the Instant Client libraries. Show us an `ls -l` listing of the Instant Client directory.

What kind of mac - Intel x86_64 or M-series?  

What's the number of the most recent timezone file on the database?

**Reply by kcjob** [reply]
I added that environment variable based on the last thing I read just before submitting my question
The error was happening before I added the variable
it's a mac intel x86)64
2 variables were set: DPI_DEBUG_LEVEL and $LD_LIBRARY_PATH
added to path: PATH=~/oracle/instantclient_19_16/:$PATH  
Mist recent TZ file: filename: timezlrg_18.dat	version: 18	con_id: 0

I have attached a file with my code
[myfile.txt](https://github.com/user-attachments/files/15796302/myfile.txt)

ls -l
-r--r--r--@ 1 kurtjob  staff       5780 Nov 10  2023 BASIC_LICENSE
-r--r--r--@ 1 kurtjob  staff       1673 Nov 10  2023 BASIC_README
-r-xr-xr-x@ 1 kurtjob  staff      69664 Nov 10  2023 adrci
-r-xr-xr-x@ 1 kurtjob  staff      88048 Nov 10  2023 genezi
lrwxrwxrwx@ 1 kurtjob  staff         20 Nov 10  2023 libclntsh.dylib -> libclntsh.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         20 Nov 10  2023 libclntsh.dylib.10.1 -> libclntsh.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         20 Nov 10  2023 libclntsh.dylib.11.1 -> libclntsh.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         20 Nov 10  2023 libclntsh.dylib.12.1 -> libclntsh.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         20 Nov 10  2023 libclntsh.dylib.18.1 -> libclntsh.dylib.19.1
-r--r--r--@ 1 kurtjob  staff   94689440 Nov 10  2023 libclntsh.dylib.19.1
-r--r--r--@ 1 kurtjob  staff    4830448 Nov 10  2023 libclntshcore.dylib.19.1
-r--r--r--@ 1 kurtjob  staff    7452176 Nov 10  2023 libnnz19.dylib
lrwxrwxrwx@ 1 kurtjob  staff         18 Nov 10  2023 libocci.dylib -> libocci.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         18 Nov 10  2023 libocci.dylib.10.1 -> libocci.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         18 Nov 10  2023 libocci.dylib.11.1 -> libocci.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         18 Nov 10  2023 libocci.dylib.12.1 -> libocci.dylib.19.1
lrwxrwxrwx@ 1 kurtjob  staff         18 Nov 10  2023 libocci.dylib.18.1 -> libocci.dylib.19.1
-r--r--r--@ 1 kurtjob  staff    1746432 Nov 10  2023 libocci.dylib.19.1
-r--r--r--@ 1 kurtjob  staff  131620000 Nov 10  2023 libociei.dylib
-r--r--r--@ 1 kurtjob  staff     183744 Nov 10  2023 libocijdbc19.dylib
-r--r--r--@ 1 kurtjob  staff     131056 Nov 10  2023 liboramysql19.dylib
drwxr-xr-x@ 3 kurtjob  staff         96 Nov  2  2023 network
-r--r--r--@ 1 kurtjob  staff    4471317 Nov 10  2023 ojdbc8.jar
-r--r--r--@ 1 kurtjob  staff    1689143 Nov 10  2023 ucp.jar
-r-xr-xr-x@ 1 kurtjob  staff     206880 Nov 10  2023 uidrvci
-r--r--r--@ 1 kurtjob  staff      41521 Nov 10  2023 xstreams.jar

**Reply by cjbj** [reply]
Can you update the DB to use the TZ file included in Instant Client, which is version 32?  You can check the version used by Instant Client by running `genezi -v`.  Or perhaps grab the old TZ file from the DB and use that with Instant Client by setting `ORA_TZFILE`, see https://python-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#oracle-environment-variables-for-python-oracledb-thick-mode. 

(@priyankanair8 that doc is slightly out of date: Instant Client 23ai and 19.18 onwards will let ORA_TZFILE be an absolute path as an easy alternative to creating the `oracore/zoneinfo` subdir, see Oracle enhancement request 31333105)

**Reply by priyankanair8** [reply]
@cjbj The doc has been updated with this information. Thank you for letting me know.

**Comment by kcjob** [comment]
Hey Christopher
I rebooted my laptop and now the program is running without errors
Thanks for your replies

**Reply by cjbj** [reply]
I'm glad it got resolved.  Thanks for letting us know.
